app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +65,138 @@ export class PersistenceManager {
    * includes a single key which is 'version' and contains the current version
    * number of the state tree.
    */
-  #metadata?: { version: number };
+  #metadata?: MetaData;
 
   #isExtensionInitialized: boolean = false;
 
-  #localStore: ExtensionStore | ReadOnlyNetworkStore;
+  #localStore: BaseStore;
+
+  #db: DB;
+
+  #open: boolean = false;
 
-  constructor({
-    localStore,
-  }: {
-    localStore: ExtensionStore | ReadOnlyNetworkStore;
-  }) {
+  constructor({ localStore }: { localStore: BaseStore }) {
     this.#localStore = localStore;
+    this.#db = new DB();
+  }
+
+  async open() {
+    if (!this.#open) {
+      await this.#db.open('metamask-vault', 1);
+      this.#open = true;
+    }
   }
 
-  setMetadata(metadata: { version: number }) {
+  setMetadata(metadata: MetaData) {
     this.#metadata = metadata;
   }
 
+  /**
+   * Sets the state in the local store.
+   *
+   * @param state - The state to set in the local store. This should be an object
+   * containing the state data to be stored.
+   * @throws Error if the state is missing or if the metadata is not set before
+   * calling this method.
+   * @throws Error if the local store is not open.
+   * @throws Error if the data persistence fails during the write operation.
+   */
   async set(state: MetaMaskStateType) {
+    await this.open();
+
     if (!state) {
       throw new Error('MetaMask - updated state is missing');
     }
-    if (!this.#metadata) {
+    const meta = this.#metadata;
+    if (!meta) {
       throw new Error('MetaMask - metadata must be set before calling "set"');
     }
-    try {
-      await this.#localStore.set({ data: state, meta: this.#metadata });
-      if (this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = false;
-      }
-    } catch (err) {
-      if (!this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = true;
-        captureException(err);
-      }
-      log.error('error setting state in local store:', err);
-    } finally {
-      this.#isExtensionInitialized = true;
-    }
+
+    await navigator.locks.request(
```

Comment:
writes are no longer atomic (since we write both to `storage.local` and to `IndexedDb` now), so we hold a lock for the duration of the write.

Reply:
Oh yikes. I'm surprised I missed that!

I can just add a polyfill for it so it isn't blocking anymore. https://www.npmjs.com/package/navigator.locks

Reply:
Tested the new polyfill in FF 95.0.1 and it works.

Regular windows don't have `navigator.locks`:
![image](https://github.com/user-attachments/assets/5e97d6ea-688c-482e-ae81-df399e4ec43f)

This branch's build of the extension still works:
![image](https://github.com/user-attachments/assets/3aa618e6-4121-4fb8-ac60-53cfe83328b9)